storage/src/main/java/tech/pegasys/artemis/storage/CombinedChainDataClient.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.storage;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static tech.pegasys.artemis.datastructures.util.BeaconStateUtil.compute_start_slot_at_epoch;
+import static tech.pegasys.artemis.datastructures.util.BeaconStateUtil.get_block_root_at_slot;
+import static tech.pegasys.artemis.util.config.Constants.SLOTS_PER_HISTORICAL_ROOT;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.artemis.datastructures.blocks.BeaconBlock;
+import tech.pegasys.artemis.datastructures.state.BeaconState;
+import tech.pegasys.artemis.datastructures.util.BeaconStateUtil;
+
+public class CombinedChainDataClient {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private static final CompletableFuture<Optional<BeaconBlock>> BLOCK_NOT_AVAILABLE =
+      completedFuture(Optional.empty());
+  private final ChainStorageClient recentChainData;
+  private final HistoricalChainData historicalChainData;
+
+  public CombinedChainDataClient(
+      final ChainStorageClient recentChainData, final HistoricalChainData historicalChainData) {
+    this.recentChainData = recentChainData;
+    this.historicalChainData = historicalChainData;
+  }
+
+  public CompletableFuture<Optional<BeaconBlock>> getBlockAtSlot(
+      final UnsignedLong slot, final Bytes32 headBlockRoot) {
+    final Store store = recentChainData.getStore();
+    if (store == null) {
+      LOG.trace("No block at slot {} because the store is not set", slot);
+      return BLOCK_NOT_AVAILABLE;
+    }
+    final BeaconState headState = store.getBlockState(headBlockRoot);
+    if (headState == null) {
+      LOG.trace("No block at slot {} because head block root {} is unknown", slot, headBlockRoot);
+      return BLOCK_NOT_AVAILABLE;
```

Comment:
It's possible that someone could be making a conservative query with a head root from the historical store.   Shouldn't we support this?

Reply:
We're only required to support head roots since the last finalised epoch:
`Clients MUST support head_block_root values since the latest finalized epoch.`
https://github.com/ethereum/eth2.0-specs/blob/v0.9.0/specs/networking/p2p-interface.md#beaconblocksbyrange

My understanding is that's what most clients will support because it allows them to check the request is on a chain they know about very quickly without having to hit their disk stores. I'm inclined to be conservative so that when we build out sync we don't assume clients will do more than what the spec requires and then have interop issues.

Reply:
Ah, cool - that makes sense.

Reply:
I realise now that this answer probably should have been on the question below though.  It would be reasonable for `CombinedChainDataClient` to support requests even if the head block root was finalised but we don't currently have any need to do it and haven't implemented a way to get a finalised state (only blocks).  We could just return the block we have finalised at the specified slot, but can't guarantee it's on the specified chain which I think is probably a worse trade off.